File: Mobile/Native/ios/objc/MeetHourSDKTest/MeetHourSDKTest/ViewController.m
//
//  ViewController.m
//  MeetHourSDKTest
//
//  Created by MeetHour, LLC.
//  Copyright © Meet Hour, LLC. All rights reserved.//

#import "ViewController.h"
#import "ConferenceViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _avatarUrl = @"https://st.depositphotos.com/1787196/1330/i/950/depositphotos_13301967-stock-photo-furry-blue-monster.jpg";
    
    UITapGestureRecognizer *gestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissKeyboard)];
       [self.view addGestureRecognizer:gestureRecognizer];
       gestureRecognizer.cancelsTouchesInView = NO;
    _isAudio = YES;
    _isVideoOn = YES;
    self.txtServerURL.text = @"https://meethour.io";
    self.room = nil;
    
    dispatch_async(dispatch_get_global_queue(0,0), ^{
        NSData * data = [[NSData alloc] initWithContentsOfURL: [NSURL URLWithString: self->_avatarUrl]];
            if ( data == nil )
                return;
            dispatch_async(dispatch_get_main_queue(), ^{
                self.imgAvatar.image = [UIImage imageWithData: data];
            });

        });
    
}
- (void)dismissKeyboard
{
     [self.view endEditing:YES];
}

- (IBAction)aMethodAudio:(UISwitch *)sender {
        self.isAudio = !self.isAudio;
}
- (IBAction)aMethodVideo:(UISwitch *)sender {
    self.isVideoOn = !self.isVideoOn;
}

- (BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender {
    // Don't present the ConferenceViewController if no room was specified.
    if ([identifier isEqualToString:@"joinConference"]) {
        self.room = self.textField.text;
        if (self.room == nil || [self.room length] == 0) {
            return NO;
        }
    }
    
    return YES;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([[segue identifier] isEqualToString:@"joinConference"]) {
        [[self view] endEditing:YES];
        // Attach the room to the new controller.
        ConferenceViewController *vc = [segue destinationViewController];
        vc.room = self.room;
        vc.pcode = @"5b40602cfea7708895781a8cad71be5b"; // Dynamically pass Encrypted Meeting Password.
        vc.serverUrl = self.txtServerURL.text;
        vc.subject = self.txtSubject.text;
        vc.displayName = self.txtDisplayName.text;
        vc.email = self.txtEmail.text;
        vc.isVideoOn = _isVideoOn;
        vc.isAudioMuted = _isAudio;
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField{
    [textField resignFirstResponder];
    return  YES;
}

@end

File: Mobile/Native/ios/swift/MeetHourSDKTest/MeetHourSDKTest/ViewController.swift
//  Created by MeetHour, LLC.
//  Copyright © Meet Hour, LLC. All rights reserved.

import UIKit
import MeetHourSDK

class ViewController: UIViewController {
    
    var isAudioMute : Bool = true
    var isVideoOn : Bool = true


    @IBOutlet weak var txtServerURL: UITextField!{
        didSet{
            txtServerURL.setLeftPaddingPoints(10)
        }
    }
    @IBOutlet weak var txtRoomName: UITextField!{
        didSet{
            txtRoomName.setLeftPaddingPoints(10)
        }
    }
    @IBOutlet weak var txtSubject: UITextField!{
        didSet{
            txtSubject.setLeftPaddingPoints(10)
        }
    }
    @IBOutlet weak var txtDisplayName: UITextField!{
        didSet{
            txtDisplayName.setLeftPaddingPoints(10)
        }
    }
    @IBOutlet weak var txtEmail: UITextField!{
        didSet{
            txtEmail.setLeftPaddingPoints(10)
        }
    }
    
    @IBOutlet weak var videoButton: UIButton?
    @IBOutlet weak var imgAvatar: UIImageView!
    
    
    let imgUrl : String = "https://st.depositphotos.com/1787196/1330/i/950/depositphotos_13301967-stock-photo-furry-blue-monster.jpg"
    
    fileprivate var pipViewCoordinator: PiPViewCoordinator?
    fileprivate var MHView: MeetHourView?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let url = URL(string: imgUrl)!
            DispatchQueue.global().async {
                if let data = try? Data(contentsOf: url) {
                    DispatchQueue.main.async {
                        self.imgAvatar.image = UIImage(data: data)
                    }
                }
            }
        let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(self.dismissKeyboard))
            view.addGestureRecognizer(gestureRecognizer)
            gestureRecognizer.cancelsTouchesInView = false
        self.txtServerURL.text = "https://meethour.io"
    }

    @objc func dismissKeyboard() {
        view.endEditing(true)
    }
    
    override func viewWillTransition(to size: CGSize,
                                     with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)

        let rect = CGRect(origin: CGPoint.zero, size: size)
        pipViewCoordinator?.resetBounds(bounds: rect)
    }


    
    @IBAction func aMethodAudio(_ sender: UISwitch) {
        isAudioMute = !isAudioMute

    }
    
    
    @IBAction func aMethodVideoOn(_ sender: UISwitch) {
        isVideoOn = !isVideoOn
    }
    
    // MARK: - Actions

    @IBAction func openMeetHour(sender: Any?) {
        
        view.endEditing(true)
        cleanUp()
        
        let info: MeetHourUserInfo = MeetHourUserInfo()
        info.displayName = self.txtDisplayName.text
        info.email = self.txtEmail.text
        info.avatar = URL(string: imgUrl)
        
        if txtRoomName.text!.count > 0{
            // create and configure Meet Hour view
            let MHView = MeetHourView()
            MHView.delegate = self
            self.MHView = MHView
            let options = MeetHourConferenceOptions.fromBuilder { (builder) in
                builder.welcomePageEnabled = false
                builder.subject = self.txtSubject.text
                builder.audioMuted = self.isAudioMute
                builder.videoMuted = self.isVideoOn
                builder.serverURL = URL(string: self.txtServerURL.text!)
                builder.userInfo = info
                builder.room = self.txtRoomName.text
                builder.pcode = "@5b40602cfea7708895781a8cad71be5b"
                builder.setFeatureFlag("ios.recording.enabled", withBoolean: true)
            }
            MHView.join(options)

            // Enable meet hour view to be a view that can be displayed
            // on top of all the things, and let the coordinator to manage
            // the view state and interactions
            pipViewCoordinator = PiPViewCoordinator(withView: MHView)
            pipViewCoordinator?.configureAsStickyView(withParentView: view)

            // animate in
            MHView.alpha = 0
            pipViewCoordinator?.show()
        }else{
            let alert = UIAlertController(title: "Alert", message: "Please enter room name.", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "Okay", style: .cancel, handler: nil))
            self.present(alert, animated: true)
        }
    }

    fileprivate func cleanUp() {
        MHView?.removeFromSuperview()
        MHView = nil
        pipViewCoordinator = nil
    }
}

extension ViewController: MeetHourViewDelegate {
    func conferenceTerminated(_ data: [AnyHashable : Any]!) {
        DispatchQueue.main.async {
            self.pipViewCoordinator?.hide() { _ in
                self.cleanUp()
            }
        }
    }

    func enterPicture(inPicture data: [AnyHashable : Any]!) {
        DispatchQueue.main.async {
            self.pipViewCoordinator?.enterPictureInPicture()
        }
    }
}

extension ViewController: UITextFieldDelegate{
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
}


extension UITextField {
    func setLeftPaddingPoints(_ amount:CGFloat){
        let paddingView = UIView(frame: CGRect(x: 0, y: 0, width: amount, height: self.frame.size.height))
        self.leftView = paddingView
        self.leftViewMode = .always
    }
    func setRightPaddingPoints(_ amount:CGFloat) {
        let paddingView = UIView(frame: CGRect(x: 0, y: 0, width: amount, height: self.frame.size.height))
        self.rightView = paddingView
        self.rightViewMode = .always
    }
}
